### ISSUE TITLE
[Found workaround] Steam broken with radeon foss drivers on unstable (16.03pre71289.7ae05ed)

### ISSUE THREAD

**Issue by pstn** [NixOS/nixpkgs#10950]
Every time I try to launch steam I only get the following message:

    mkdir: cannot create directory ‘/tmp’: File exists
    cp: cannot create regular file ‘/home/philipp/.local/share/Steam/bootstrap.tar.xz’: Permission denied
    Running Steam on nixos 16.03pre71289.7ae05ed 64-bit
    STEAM_RUNTIME has been set by the user to: /steamrt
    Installing breakpad exception handler for appid(steam)/version(0_client)
    libGL error: unable to load driver: radeonsi_dri.so
    libGL error: driver pointer missing
    libGL error: failed to load driver: radeonsi
    libGL error: unable to load driver: swrast_dri.so
    libGL error: failed to load driver: swrast

* nixos-version: 16.03pre71289.7ae05ed
* I already tried to remove ~/.local/share/Steam and it didn't help
* Excerpt of my nix config:
hardware-configuration.nix

    hardware = {
     opengl.driSupport = true;
     pulseaudio.enable = true;
     opengl.driSupport32Bit = true;
     pulseaudio.support32Bit = true;
    };

* [xorg.log](http://pastebin.ca/3246838)
* [glxinfo](http://pastebin.ca/3246840)

Thank you for your help, I will happily provide any more information you might need.

**Comment by abbradar** [comment]
Ouch, this is a new incarnation of https://github.com/NixOS/nixpkgs/issues/9003. I'll look into this, thanks for the reporting!

**Comment by vcunat** [comment]
When did it last work for you? (on the same machine/configuration) For example, I have nonstandard mount setup of $HOME, which results into missing vital stuff in the chroot by default (I add a mount by hand into the chroot).

**Comment by pstn** [comment]
The setup never worked since this is a new install (I'm new to nixos). 
My home is pretty standard:

    /dev/mapper/lvm-philipphome on /home/philipp type ext4 (rw,relatime,data=ordered)

**Comment by abbradar** [comment]
Hmm, that's strange -- we have `libstdc++` overridden by our version, which should fix the bug. I suspect Steam uses its own `libstdc++` anyway -- I'll investigate more for a chance to avoid using nasty `LD_PRELOAD` again.

**Comment by abbradar** [comment]
Looks like there's sadly no other option. Can you please test if https://github.com/abbradar/nixpkgs/commit/0e7cc21b33f3406cebf0851e6a11761b0864639b fixes the problem for you?

1. Make `test.nix` file with contents like this in the root of `nixpkgs`:
```
{ pkgs ? import ./ { } }:
pkgs.steam.override { workaroundLibstdc = true; }
```
2. Run `nix-build test.nix`
3. Run `result/bin/steam` and test.

**Comment by pstn** [comment]
ι tried to run your test but I can't get it to work. If I follow your guide as it is, I just get 
`error: syntax error, unexpected '/', expecting ID or OR_KW or DOLLAR_CURLY or '"', at /home/philipp/workspace/nixpkgs/test.nix:1:18`
If I change `./` to the directory I cloned from github I get 

    error: anonymous function at /home/philipp/workspace/nixpkgs/pkgs/games/steam/default.nix:1:1 called with unexpected argument ‘workaroundLibstdc’, at /home/philipp/workspace/nixpkg /lib/customisation.nix:58:12

I also ran `LD_PRELOAD='/steamrt/amd64/lib/libstdc++.so.6:/steamrt/i386/lib/libstdc++.so.6' steam`
but that only returns

    ERROR: ld.so: object '/steamrt/amd64/lib/libstdc++.so.6' from LD_PRELOAD cannot be preloaded (cannot open shared object file): ignored.
    ERROR: ld.so: object '/steamrt/i386/lib/libstdc++.so.6' from LD_PRELOAD cannot be preloaded (cannot open shared object file): ignored.
    ERROR: ld.so: object '/steamrt/amd64/lib/libstdc++.so.6' from LD_PRELOAD cannot be preloaded (cannot open shared object file): ignored.
    ERROR: ld.so: object '/steamrt/i386/lib/libstdc++.so.6' from LD_PRELOAD cannot be preloaded (cannot open shared object file): ignored.
    mkdir: cannot create directory ‘/tmp’: File exists
    cp: cannot create regular file ‘/home/philipp/.local/share/Steam/bootstrap.tar.xz’: Permission denied
    Running Steam on nixos 16.03pre71289.7ae05ed 64-bit
    STEAM_RUNTIME has been set by the user to: /steamrt
    Installing breakpad exception handler for appid(steam)/version(0_client)
    libGL error: unable to load driver: radeonsi_dri.so
    libGL error: driver pointer missing
    libGL error: failed to load driver: radeonsi
    libGL error: unable to load driver: swrast_dri.so
    libGL error: failed to load driver: swrast

**Comment by abbradar** [comment]
Ouch, I'm sorry, this should have been `./.`, not `./`. Anyway, have you cherry-picked my patch? It's strange it doesn't have `workaroundLibstdc` argument, I added it with the patch.

**Comment by pstn** [comment]
I did cherry pick your patch, there was a merge conflict but [I resolved it](http://pastebin.ca/3255039). Maybe I made a mistake?

**Comment by abbradar** [comment]
Hmm, the resulting chrootenv looks veery old, I think it was like that somewhere in the spring and definitely before 15.09. Try this branch: https://github.com/abbradar/nixpkgs/tree/steam-stdlibc . It's on top of latest `unstable` channel.

**Comment by jgillich** [comment]
Also seeing this with intel:
```
cp: cannot create regular file ‘/home/jakob/.local/share/Steam/bootstrap.tar.xz’: Permission denied
Running Steam on nixos 16.03.git.c828102 64-bit
STEAM_RUNTIME has been set by the user to: /steamrt
Installing breakpad exception handler for appid(steam)/version(0_client)
libGL error: failed to open drm device: Permission denied
libGL error: failed to load driver: i965
libGL error: unable to load driver: swrast_dri.so
libGL error: failed to load driver: swrast
```

**Comment by abbradar** [comment]
Does anyone still see this? If yes, on what branch?

**Comment by pstn** [comment]
I do still see it on `` https://nixos.org/channels/nixos-unstable ``

**Comment by abbradar** [comment]
Can you please try [https://github.com/abbradar/nixpkgs/commit/dc965cab44ddd4d04017f8ed01bcd5f537de0e4b](this patch) and see if at least some games are fixed for you? Use the old guide above.

**Comment by pstn** [comment]
Unfortunately [it doesn't](https://gist.github.com/anonymous/f405b489654746901f3e).

**Comment by abbradar** [comment]
Hm, there is one leftover `libstdc++` that I see -- try https://github.com/abbradar/nixpkgs/commit/c9a98f13deb73b0dbc0fb486427d30d40cdc87b8, please.

**Comment by pstn** [comment]
That also [didn't work](https://gist.github.com/3958497aeda9b4de7f6c).

**Comment by abbradar** [comment]
https://github.com/abbradar/nixpkgs/commit/ee13ba8174864dd88a66eb0cb2401cf0d43987b9 this one, next!

**Comment by pstn** [comment]
That did it! Should I revert the other patches to look whether this fix alone is enough or would they be applied anyway?

**Comment by abbradar** [comment]
Yes, please. Also please test what games do work for you -- some most certainly wouldn't.

**Comment by pstn** [comment]
* The last patch alone fixes the issue with steam not starting.
* VVVVVV runs fine.
* Steam crashes when I try to add another library folder with a segfault.
* I don't have any sound because pulseaudio denies a connection but I think that's another unrelated problem
* I will try more games as I download them.

**Comment by abbradar** [comment]
Revised patchset at https://github.com/NixOS/nixpkgs/pull/12404

**Comment by abbradar** [comment]
Sadly I don't have any statistics on games which actually work with new `libstdc++` -- I think someone reported DotA 2 as broken before, but I don't have the game or a Radeon card to test myself.

**Comment by abbradar** [comment]
Let's keep this bug open for discovering possible new workarounds, tests and Steam updates.

**Comment by pstn** [comment]
Most games seem to run. I have been able to run and get into actual gameplay of
* VVVVVV
* The Stanley Parable (Source Engine)
* Kerbal Space Program (Unity Engine)
* Xenonauts
* DOTA2 with Reborn Beta (Source 2)
I couldn't find any games that do not run so far.

**Comment by abbradar** [comment]
Please try [this branch](https://github.com/abbradar/nixpkgs/commits/mesa-steam) with this `test.nix` (as in old guide):
```
{ pkgs ? import ./ { } }:
pkgs.steam.override { withMesa = true; }
```
This will have a non-trivial amount of rebuilding, but should be the proper fix of your problem.

**Comment by pstn** [comment]
I tried your branch bit it didn't work. Sorry for taking so long.

**Comment by abbradar** [comment]
Hm, what would happen if you clean LD_LIBRARY_PATH? Run `LD_LIBRARY_PATH= steam`.

**Comment by pstn** [comment]
That doesn't change anything.

**Comment by abbradar** [comment]
Oh, that was because I actually set `LD_LIBRARY_PATH` somewhere inside `chrootenv` itself. Pull from the branch and try again, please.

**Comment by pstn** [comment]
That build also failed in the same way.

**Comment by abbradar** [comment]
Very strange. I'm out for a bit, can you please make two changes in `pkgs/games/steam/chrootenv.nix`in my branch:
* Change `steam` to `bash` in `runScript`;
* Add `glxinfo` to `targetPackages`. 
Then run and call:
* `echo $LD_LIBRARY_PATH
* `glxinfo`
And show me the results?

**Comment by pstn** [comment]
[glxinfo](https://gist.github.com/anonymous/d283f6feb112207e474a)

**Comment by abbradar** [comment]
More and more interesting. What's in LD_LIBRARY_PATH?

**Comment by pstn** [comment]
Sorry, I forgot to say that it's empty.

**Comment by abbradar** [comment]
...and Steam fails nevertheless, huh? 
`find /steamrt -name \*stdc\*`
`find /usr - name \*stdc\*`
Steam fails with the same `cannot load driver`, correct?

**Comment by pstn** [comment]
The exact errors are still on steam launch are
```
libGL error: unable to load driver: radeonsi_dri.so
libGL error: driver pointer missing
libGL error: failed to load driver: radeonsi
libGL error: unable to load driver: swrast_dri.so
libGL error: failed to load driver: swrast
```
But I guess this is what you meant.
Both finds don't find anything inside the steam-chrootenv.

**Comment by pstn** [comment]
Follow up for new instance of this error:
glxgears is only 32Bit with 
     `nix-build -E 'with import ./. {}; pkgsi686Linux.steamPackages.steam-runtime-wrapped.override {newStdcpp = true; }'`
and 64Bit with
`nix-build -E 'with import ./. {}; steamPackages.steam-runtime-wrapped.override { newStdcpp =  true; }'`
Both run fine.

**Comment by mirh** [comment]
What does `LIBGL_DEBUG=verbose steam` say[?](https://bbs.archlinux.org/viewtopic.php?id=193802#p1503924)
In my case it complains about

    libGL: dlopen /usr/lib32/xorg/modules/dri/r600_dri.so failed (/home/user/.local/share/Steam/ubuntu12_32/steam-runtime/i386/usr/lib/i386-linux-gnu/libstdc++.so.6: version `CXXABI_1.3.8' not found (required by /usr/lib32/xorg/modules/dri/r600_dri.so))
which basically resolves down to [this problem](http://stackoverflow.com/questions/23494103/version-cxxabi-1-3-8-not-found-required-by). Normally you would be able to override this kind of error with LD_LIBRARY_PATH variable as you say above, but steam.sh (line 702) overrides.. any previous override. So you are back to start. 

Editing the start script or deleting libs from steam folder seemed unorthodox. 
So (once taken into account of [another](https://manerosss.wordpress.com/2016/05/16/how-to-fix-steam-not-running-with-open-source-drivers-on-arch-linux/) error) I fixed it with `export LD_PRELOAD=/usr/lib32/libstdc++.so.6:/usr/lib32/libgpg-error.so.0`

p.s: adjust paths for Nix, should mine from Arch not compute :s

**Comment by pstn** [comment]
I already resolved the issue with @abbradar should have said something about that, sorry. We are keeping this thread open because this is a recurring error that comes back due to different upgrades or changes of the system and we need to resolve it again.

**Comment by vcunat** [comment]
Then maybe the first comment could be edited to be more informative...

**Comment by rardiol** [comment]
I'm getting the following error. Is this the same issue?

cp: cannot create regular file '/home/user/.local/share/Steam/bootstrap.tar.xz': Permission denied
Running Steam on nixos 16.09pre87332.715e01c 64-bit
STEAM_RUNTIME has been set by the user to: /steamrt
Installing breakpad exception handler for appid(steam)/version(0)
libGL error: unable to load driver: nouveau_dri.so
libGL error: driver pointer missing
libGL error: failed to load driver: nouveau
libGL error: unable to load driver: swrast_dri.so
libGL error: failed to load driver: swrast

**Comment by pstn** [comment]
It's probably a similar issue. Here are a few questions to get things started:
* What version of nixos are you using?
* What graphics card are you using?
* What drivers are you using?
* Do you have 
 ```nixpkgs.config.steam = pkgs.steam.override {
            newStdcpp = true;
    };``` set somewhere in your nix-config?

**Comment by rardiol** [comment]
16.09pre87332.715e01c (Flounder)
01:00.0 VGA compatible controller: NVIDIA Corporation GT218M [GeForce 310M] (rev a2)
nouveau
No. should I?

**Comment by pstn** [comment]
Yes, please try that. Never mesa needs another libcpp than what is in the steam runtime. This should fix it.

**Comment by rardiol** [comment]
That seems to work thanks.

**Comment by Mounium** [comment]
https://gist.github.com/Mounium/98eb1587d066210a4fe9b3f735260db9

May it be the related issue (except for this time not at starting steam, but starting a game _(borderlands 2)_)? Steam starts up fine with any of the following arguments being set to true. I tried newStdcpp, nativeOnly and runtimeOnly as well, all three display the linked error messages, also both 16.03 version and unstable. I have an AMD R9 380 (Tonga).

**Comment by pstn** [comment]
I own this specific game and could reproduce the error. 
My first guess is that it's not directly related but that there is something missing in our runtime.

**Comment by Shados** [comment]
@abbradar @pstn I'm seeing the same error when using the radeon driver, that is:
```
libGL error: unable to load driver: radeonsi_dri.so
libGL error: driver pointer missing
libGL error: failed to load driver: radeonsi
libGL error: unable to load driver: swrast_dri.so
libGL error: failed to load driver: swrast
```

Setting `nixpkgs.config.steam = pkgs.steam.override { newStdcpp = true; };` does not help.

**Comment by abbradar** [comment]
To install overridden Steam you need to either use
nixpkgs.config.packageOverrides or use it directly in
environment.systemPackages:

environment.systemPackages = [(pkgs.steam.override { newStdcpp = true; })];

On August 12, 2016 5:06:02 PM GMT+03:00, Alexei Robyn
<notifications@github.com> wrote:

    @abbradar <https://github.com/abbradar> @pstn
    <https://github.com/pstn> I'm seeing the same error when using the
    radeon driver, that is:

    |libGL error: unable to load driver: radeonsi_dri.so libGL error:
    driver pointer missing libGL error: failed to load driver: radeonsi
    libGL error: unable to load driver: swrast_dri.so libGL error:
    failed to load driver: swrast |

    Setting |nixpkgs.config.steam = pkgs.steam.override { newStdcpp =
    true; };| does not help.

    —
    You are receiving this because you were mentioned.
    Reply to this email directly, view it on GitHub
    <https://github.com/NixOS/nixpkgs/issues/10950#issuecomment-239455244>,
    or mute the thread

**Comment by cleverca22** [comment]
after messing with steam for a few hours, i managed to get it to run with https://gist.github.com/cleverca22/2fcaa5852bf13d80aa7b6bb3b285be9f

this proves that its not the opengl drivers lacking a feature like i first suspected, and it now feels like steam.sh is breaking the opengl drivers, i've begun editing mesa to make it more debugable

**Comment by cleverca22** [comment]
heh
i was told on irc that newStdcpp didnt work, but after hacking up my mesa source to show better errors i got
``libGL error: dlopen /run/opengl-driver-32/lib/dri/radeonsi_dri.so failed (/steamrt/i386/usr/lib/i386-linux-gnu/libstdc++.so.6: version `GLIBCXX_3.4.21' not found (required by /run/opengl-driver-32/lib/dri/radeonsi_dri.so))``

so i tried `nix-build -E 'with import <nixpkgs> {}; steam.override { newStdcpp = true; }' && ./result/bin/steam` and now it works just fine

though dosbox went full-screen mirrored, and failed to restore the monitor layout upon exit

**Comment by Shados** [comment]
@abbradar That gets steam itself running, but I am seeing issues with all the 3D games I tried:
- Dota 2 comes up with 'Failed to create an OpenGL context. Your graphics card must support at least OpenGL v3.1.'
- Dead Island crashes, stdout shows `CLOG: [CGame::AppWindowInit] Could not create GL context: GLXBadFBConfig[OpenGL] "Context" created.TTL_ASSERT:`
- Distance silently crashes

2D games I tested launched OK (Hammerwatch, Factorio, Binding of Isaac: Rebirth). glxgears runs fine, [here](http://sprunge.us/UXNf)'s the output of glxinfo.

**Comment by abbradar** [comment]
@Shados Sadly I can't test this myself because I don't own an ATI card. I feel that we just need to override yet another library, but I don't know which one...

**Comment by pstn** [comment]
I have a radeon card and dota launches without any problems on the latest master, so it's probably not a library, because those should be the same for us two, I guess.

**Comment by neeasade** [comment]
> Dota 2 comes up with 'Failed to create an OpenGL context. Your graphics card must support at least OpenGL v3.1.'

can we re-open this or should this be a new issue? I see the same on 17.03.

**Comment by vcunat** [comment]
Could this be due to #24116? (only fixed recently)

**Comment by vcunat** [comment]
For reference, on nouveau steam only runs with `newStdcpp = true`, for me at least.

I tried steam from 17.03 and unstable+ with NixOS on unstable+.  Binary nvidia drivers don't seem to have problems with the default.

**Comment by wtok** [comment]
I just installed NixOS and ran into this issue. It turned out I had installed Steam via `nix-env -i steam` as a regular user and putting override settings in my system-wide config had no effect. 

So I needed to have Steam listed as a system-level package, by including it in `/etc/nixos/configuration.nix`, with the newStdcpp override, by adding this to that file:

```
environment.systemPackages = with pkgs; [ 
    # whatever other packages ...
    (steam.override { newStdcpp = true; })
];
```

You might need to merge that with your current `environment.systemPackages` setting if it's already in there.

Hopefully that helps someone in future. In all my searching for solutions to this issue, I never saw it mentioned that installing via `nix-env` as a regular user might lead to this issue, so there it is.

If there's some way to set the override such that it'll affect a user-installed Steam, that might be a better solution.

**Comment by aanderse** [comment]
wtok - thank you, this solved my issue!

**Comment by RodrigoLeiteF** [comment]
Hey. I'm getting this error when trying to run League of Legends on the Radeon RX 5700 XT. Can we take another look, please?

**Comment by pstn** [comment]
Afaik League of Legends is neither on steam, nor natively on linux. How are you running it?
